compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -0,0 +1,402 @@
+        return super.visitConstructorCall(expression)
+    }
+
+    override fun visitDelegatingConstructorCall(expression: IrDelegatingConstructorCall): IrExpression {
+        val constructor = expression.symbol.owner
+        if (constructor.isPrimaryInlineClassConstructor)
+            return visitPrimaryInlineClassConstructorCall(expression, constructor)
+        return super.visitDelegatingConstructorCall(expression)
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        val function = expression.symbol.owner
+        if (!function.isInlineClassFieldGetter)
+            return super.visitCall(expression)
+        val arg = expression.dispatchReceiver!!.transform(this, null)
+        return coerceInlineClasses(arg, function.dispatchReceiverParameter!!.type, expression.type)
+    }
+
+    override fun visitGetField(expression: IrGetField): IrExpression {
+        val field = expression.symbol.owner
+        val parent = field.parent
+        if (parent is IrClass && parent.isInline && field.name == parent.inlineClassFieldName) {
+            val receiver = expression.receiver!!.transform(this, null)
+            return coerceInlineClasses(receiver, receiver.type, field.type)
+        }
+        return super.visitGetField(expression)
+    }
+
+    override fun visitReturn(expression: IrReturn): IrExpression {
+        expression.returnTargetSymbol.owner.safeAs<IrFunction>()?.let { target ->
+            manager.getReplacementFunction(target)?.let {
+                return context.createIrBuilder(it.function.symbol).irReturn(
+                    expression.value.transform(this, null)
+                )
+            }
+        }
+        return super.visitReturn(expression)
+    }
+
+    private fun visitStatementContainer(container: IrStatementContainer) {
+        container.statements.transformFlat { statement ->
+            if (statement is IrFunction)
+                transformFunctionFlat(statement)
+            else
+                listOf(statement.transform(this, null))
+        }
+    }
+
+    override fun visitContainerExpression(expression: IrContainerExpression): IrExpression {
+        visitStatementContainer(expression)
+        return expression
+    }
+
+    override fun visitBlockBody(body: IrBlockBody): IrBody {
+        visitStatementContainer(body)
+        return body
+    }
+
+    override fun visitLocalDelegatedProperty(declaration: IrLocalDelegatedProperty): IrStatement {
+        declaration.delegate.transformChildrenVoid()
+        declaration.getter = transformFunctionFlat(declaration.getter)[0] as IrFunction
+        declaration.setter?.let { setter ->
+            declaration.setter = transformFunctionFlat(setter)[0] as IrFunction
+        }
+        return declaration
+    }
+
+    private fun buildPrimaryInlineClassConstructor(irClass: IrClass, irConstructor: IrConstructor) {
+        irClass.declarations += buildConstructor {
```

Comment:
Please consider using `addConstructor`/`addFunction`/`addValueParameter`/... family of functions instead of `buildConstructor` et al, throughout the file. The advantage is that they automatically set `parent` and add the new declaration to parent's `declarations`

Reply:
I've changed it here, but cannot change it in `build(Un)BoxFunction` or `InlineClassManager.kt` since these functions potentially run before the lowering and for classes in other files. In this case we don't want to add the functions to the class (yet).